src/rust/engine/src/externs/fs.rs

```diff
@@ -229,6 +242,18 @@ impl PySnapshot {
     )
   }
 
+  #[getter]
+  fn symlinks<'py>(&self, py: Python<'py>) -> &'py PyTuple {
```

Comment:
So, thinking about this more (in the context of being [symlink oblivious](https://github.com/pantsbuild/pants/issues/10271#issuecomment-1247086217)).

Currently `Snapshot` is symlink oblivious, in that users don't have to care about whether something is a symlink... they can view whatever was behind the symlink as a `Directory` or `File`. It's possible that even if `DigestTrie`/`Digest` have become symlink aware, that `Snapshot` should be symlink oblivious, for ease of use.

That's definitely subtle though, because it would probably only be `Snapshot` and `DigestContents` which would want to be symlink oblivious. `CreateDigest` and `DigestEntries` need to maintain their symmetry, and they should definitely gain the ability to manipulate symlinks in this PR.

Reply:
Just to make sure I understand. The `Snapshot.digest` would still contain the symlink information, but the properties `files` and `dirs` are symlink oblivious?

Reply:
It doesn't need to be done in this PR but we could use a similar approach to what is done for `Process` -> `FallibleProcessResult` -> `ProcessResult`. That is, introduce a new `RawSnapshot` type (name TBD) that contains symlinks (along with files and directories) and converts to `Snapshot` by being symlink oblivious via new core rule.

Reply:
> Just to make sure I understand. The `Snapshot.digest` would still contain the symlink information, but the properties `files` and `dirs` are symlink oblivious?

That's right.

> It doesn't need to be done in this PR but we could use a similar approach to what is done for `Process` -> `FallibleProcessResult` -> `ProcessResult`. That is, introduce a new `RawSnapshot` type (name TBD) that contains symlinks (along with files and directories) and converts to `Snapshot` by being symlink oblivious via a new core rule that does the conversion.

We could, but I think that `Snapshot` vs `DigestEntries` might already represent sufficiently high/low level APIs to make having two kinds of snapshot unnecessary?

Reply:
> We could, but I think that Snapshot vs DigestEntries might already represent sufficiently high/low level APIs to make having two kinds of snapshot unnecessary?

Maybe? Do we see any users of `Snapshot` wanting to be symlink aware?

Reply:
> Maybe? Do we see any users of `Snapshot` that would want to be symlink aware?

I can't really imagine what the use for it would be. Just knowing that a symlink exists (as a string name via `fn links() -> Vec<String>`) isn't really enough to do anything with it... you need to know the type.

Reply:
`links` could return a list of (SOURCE, TARGET) tuples. So I agree with your point about this API being sufficient for now. If we don't see an immediate need for `Snapshot` users to need such information, then easy enough to defer on providing a solution given `DigestEntries` can be used to get the symlinks for any callers that need it.

Reply:
> So I agree with your point about this API being sufficient for now.

To be clear: this API is not quite right at the moment: see https://github.com/pantsbuild/pants/pull/16844#discussion_r971274624

The `files`/`directories` methods of `PySnapshot` would need to be `SymlinkBehavior::Oblivious`, in order to return the files/dirs behind symlinks.